Sources/Tracing/Docs.docc/Guides/InstrumentYourLibrary.md

```diff
@@ -0,0 +1,294 @@
+  // ...
+  
+  // 2) restore baggage, using a task-local:
+  await Baggage.withValue(baggage) {
+    // 3) start span, using contextual baggage (which may contain trace-ids already):
+    await InstrumentationSystem.tracer.withSpan("\(request.path)") { span in 
+      // 3.1) Set useful attributes:on the span:
+      span.attributes["http.method"] = request.method
+      // ... 
+      // See also: Open Telemetry typed attributes in swift-distributed-tracing-extras
+                                                                    
+      // 4) user code will have the apropriate Span baggage restored:
+      await userCode(request)
+    }
+  }
+}
+```
+
+This is introducing multiple layers of nesting, and we have un-necessarily restored, picked-up, and restored the baggage again. In order to avoid this duplicate work, it is beneficial to use the ``TracerProtocol/withSpan(_:baggage:ofKind:_:)-1w7i6`` overload, which also accepts a `Baggage` as parameter, rather than picking it up from the task-local value:
+
+```swift
+// BETTER
+func handler(request: HTTPRequest) async { 
+  // 1) extract trace information into baggage: 
+  InstrumentationSystem.instrument.extract(..., into: &baggage, ...)
+
+  // 2) start span, passing the freshly extracted baggage explicitly:
+  await InstrumentationSystem.tracer.withSpan("\(request.path)", baggage: baggage) { span in 
+    // ... 
+  }
+}
+```
+
+This method will only restore the baggage once, after the tracer has had a chance to decide if this execution will be traced, and if so, setting its own trace and span identifiers. This way only one task-local access (set) is performed in this version of the code, which is preferable to the set/read/set that was performed previously.
+
+#### Manual Span Lifetime Management
+
+While the ``TracerProtocol/withSpan(_:ofKind:_:)-11n3y`` API is preferable in most situations, it may not be possible to use when the lifetime of a span only terminates in yet another callback API. In such situations, it may be impossible to "wrap" the entire piece of code that would logically represent "the span" using a `withSpan(...) { ... }` call.
+
+In such situations you can resort to using the ``TracerProtocol/startSpan(_:baggage:ofKind:)`` and ``Span/end()`` APIs explicitly. Those APIs can then be used like this:
+
+```swift
+// Callback heavy APIs may need to store and manage spans manually:
+var span: Span? 
+
+func startHandline(request: HTTPRequest) {
+  self.span = InstrumentationSystem.tracer.startSpan("\(request.path)")
+  
+  userCode.handle(request)
+}
+
+func finishHandling(request: HTTPRequest, response: HTTPResponse) {
+  // always end spans (!)
+  span?.end() // ends and flushes the span
+}
+```
+
+It is very important to _always_ end spans that are started, as otherwise resources they use may keep accumulating and lead to memory leaks, or worse issues in tracing backends dependin on their implementation.
```

Comment:
```suggestion
It is very important to _always_ end spans that have been started, as otherwise resources they use may keep accumulating and lead to memory leaks, or worse issues in tracing backends depending on their implementation.
```